File: crates/hyper_rhi/src/pipeline_layout.rs
//
// Copyright (c) 2024, SkillerRaptor
//
// SPDX-License-Identifier: MIT
//

use std::fmt::Debug;

use downcast_rs::Downcast;

#[derive(Clone, Debug)]
pub struct PipelineLayoutDescriptor<'a> {
    pub label: Option<&'a str>,
    pub push_constants_size: usize,
}

pub trait PipelineLayout: Debug + Downcast {
    fn push_constants_size(&self) -> usize;
}

downcast_rs::impl_downcast!(PipelineLayout);

File: hyper_engine/src/rhi/pipeline_layout.cpp
/*
 * Copyright (c) 2025-present, SkillerRaptor
 *
 * SPDX-License-Identifier: MIT
 */

#include "hyper_rhi/pipeline_layout.hpp"

namespace hyper_engine
{
    PipelineLayout::PipelineLayout(const PipelineLayoutDescriptor &descriptor)
        : m_label(descriptor.label)
        , m_push_constant_size(descriptor.push_constant_size)
    {
    }

    std::string_view PipelineLayout::label() const
    {
        return m_label;
    }

    uint32_t PipelineLayout::push_constant_size() const
    {
        return m_push_constant_size;
    }
} // namespace hyper_engine